File: api-go/cmd/server/servidor.go
// Command server levanta la API Go de factorización QR.
package main

import (
	"context"
	"log/slog"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/gofiber/fiber/v3"

	"github.com/socius/interseguro-challenge/api-go/internal/api"
	"github.com/socius/interseguro-challenge/api-go/internal/config"
)

func main() {
	// Log estructurado en JSON: es lo que esperan los agregadores de las
	// plataformas cloud (Cloud Logging, CloudWatch) para indexar los campos.
	registrador := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo}))
	slog.SetDefault(registrador)

	configuracion, err := config.Cargar()
	if err != nil {
		registrador.Error("configuración inválida", slog.Any("error", err))
		os.Exit(1)
	}

	aplicacion := api.NuevaAplicacion(configuracion, registrador)

	// Se escucha en 0.0.0.0 para ser alcanzable desde fuera del contenedor.
	direccion := "0.0.0.0:" + configuracion.Puerto

	// El servidor corre en su propia goroutine para que el hilo principal
	// quede libre esperando la señal de apagado.
	errorServidor := make(chan error, 1)
	go func() {
		registrador.Info("servidor iniciado",
			slog.String("addr", direccion),
			slog.String("version", api.Version),
			slog.String("statsApi", configuracion.URLAPIEstadisticas),
		)
		errorServidor <- aplicacion.Listen(direccion, fiber.ListenConfig{
			// El banner ASCII de arranque ensuciaría el log estructurado: la
			// misma información ya sale como evento JSON justo arriba.
			DisableStartupMessage: true,
		})
	}()

	// SIGTERM es la señal que envía Docker (y Kubernetes) al detener un
	// contenedor; SIGINT llega con Ctrl+C en desarrollo.
	apagado := make(chan os.Signal, 1)
	signal.Notify(apagado, os.Interrupt, syscall.SIGTERM)

	select {
	case err := <-errorServidor:
		if err != nil {
			registrador.Error("el servidor se detuvo con error", slog.Any("error", err))
			os.Exit(1)
		}

	case senal := <-apagado:
		registrador.Info("apagado solicitado, drenando conexiones", slog.String("signal", senal.String()))

		// Apagado ordenado: se da margen a los requests en curso para terminar
		// antes de cerrar. Sin esto, un despliegue cortaría respuestas a medio
		// camino y el cliente vería errores de red sin causa aparente.
		contexto, cancelar := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancelar()

		if err := aplicacion.ShutdownWithContext(contexto); err != nil {
			registrador.Error("el apagado ordenado no completó", slog.Any("error", err))
			os.Exit(1)
		}
		registrador.Info("servidor detenido")
	}
}

File: api-node/src/servidor.ts
/** Punto de entrada: levanta la API de estadísticas. */

import { crearAplicacion, crearRegistrador } from './aplicacion.js';
import { ErrorConfiguracion, cargarConfiguracion } from './configuracion.js';

function main(): void {
  let configuracion;
  try {
    configuracion = cargarConfiguracion();
  } catch (error) {
    if (error instanceof ErrorConfiguracion) {
      // Todavía no hay logger configurado, así que el error va a stderr crudo.
      process.stderr.write(`configuración inválida: ${error.message}\n`);
      process.exit(1);
    }
    throw error;
  }

  const registrador = crearRegistrador(configuracion);
  const aplicacion = crearAplicacion(configuracion, registrador);

  // 0.0.0.0 en lugar de localhost: dentro de un contenedor, escuchar solo en la
  // interfaz de loopback lo haría inalcanzable desde fuera.
  const servidor = aplicacion.listen(configuracion.puerto, '0.0.0.0', () => {
    registrador.info({ puerto: configuracion.puerto }, 'servidor iniciado');
  });

  // Los valores predeterminados de Node pueden cambiar entre versiones. Al
  // declararlos se evita que clientes lentos retengan conexiones sin límite y
  // el comportamiento queda documentado junto al servidor.
  servidor.requestTimeout = 30_000;
  servidor.headersTimeout = 10_000;
  servidor.keepAliveTimeout = 5_000;

  // Apagado ordenado: se deja de aceptar conexiones y se espera a que terminen
  // los requests en curso. Sin esto, un despliegue cortaría respuestas a medio
  // camino y el cliente vería errores de red sin causa aparente.
  let apagadoIniciado = false;
  const apagar = (senal: string): void => {
    if (apagadoIniciado) return;
    apagadoIniciado = true;
    registrador.info({ signal: senal }, 'apagado solicitado, drenando conexiones');

    // Red de seguridad: si alguna conexión no cierra, no se puede quedar
    // colgado indefinidamente bloqueando el despliegue.
    const salidaForzada = setTimeout(() => {
      registrador.error('el apagado ordenado no completó a tiempo, forzando la salida');
      process.exit(1);
    }, 10_000);
    salidaForzada.unref();

    servidor.close((error) => {
      if (error) {
        registrador.error({ err: error }, 'error al cerrar el servidor');
        process.exit(1);
      }
      registrador.info('servidor detenido');
      process.exit(0);
    });
  };

  // SIGTERM es la señal que envía Docker al detener un contenedor;
  // SIGINT llega con Ctrl+C en desarrollo.
  process.once('SIGTERM', () => apagar('SIGTERM'));
  process.once('SIGINT', () => apagar('SIGINT'));
}

main();
